.github/workflows/ci.yml

```diff
@@ -0,0 +1,140 @@
+name: CI Workflow
+on:
+  push:
+    branches:
+      - main
+    tags:
+      - rel/**
+  pull_request:
+    branches:
+      - main
+
+permissions: read-all
+
+jobs:
+  lint:
+    name: Linting
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@v4
+
+      - name: Cache pre-commit
+        uses: actions/cache@v4
+        with:
+          path: ${{ runner.home }}/.cache/pre-commit
+          key: ${{ runner.os }}-pre-commit-${{ hashFiles('.pre-commit-config.yaml') }}
+          restore-keys: ${{ runner.os }}-pre-commit-
+
+      - name: Setup Python
+        uses: actions/setup-python@v5
+        with:
+          python-version: '3.10.14'
+
+      - name: Install dev dependencies
+        run: pip install -r requirements-dev.txt
+
+      - name: Run pre-commit linting
+        run: pre-commit run --all-files
+
+  system-tests:
+    name: System Tests
+    runs-on: windows-latest
+    needs: [tests,lint]
+    if: github.event_name != 'pull_request' && (github.ref == 'refs/heads/main' || startsWith(github.ref, 'refs/tags/rel/'))
+    permissions:
+      contents: write
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@v4
+
+      - name: Setup Python
+        uses: actions/setup-python@v5
+        with:
+          python-version: '3.10.8'
+
+      - name: Install test dependencies
+        run: |
+          pip install -r requirements-test.txt
+          pip install tox
+          pip install --upgrade coverage pytest pytest-cov
```

Comment:
rather than running `pip install` / `pip install --upgrade` we should maybe just add the up-to-date depedencies to the `requirements-test.txt`. This could otherwise install dynamic versions and break without code changes.